packages/server/graphql/public/queries/search.ts

```diff
@@ -0,0 +1,382 @@
+  })
+
+  const semanticMap = new Map<string, number>()
+  const semanticScore = new Map<string, number>()
+  const semanticSnippets = new Map<string, string[]>()
+
+  semanticResults.forEach((r, i) => {
+    const id = String(r.metadataId)
+    if (!semanticMap.has(id)) {
+      semanticMap.set(id, i + 1)
+      semanticScore.set(id, r.score)
+    }
+    const snippets = semanticSnippets.get(id) || []
+    if (snippets.length < 3) snippets.push(r.embedText || '')
+    semanticSnippets.set(id, snippets)
+  })
+
+  // Calculate RRF for Documents
+  const rrfScores = calculateRRF([lexicalMap, semanticMap], k, [keywordWeight, vectorWeight])
+
+  const offset = rawOffset || 0
+
+  // Sort and Paginate
+  const topIds = Array.from(rrfScores.entries())
+    .sort((a, b) => b[1] - a[1])
+    .slice(offset, offset + limit)
+    .map(([id]) => Number(id))
+
+  if (topIds.length === 0) {
+    return {results: [], totalCount: 0}
+  }
+
+  // Fetch Details
+  const metadata = await pg
+    .selectFrom('EmbeddingsMetadata as m')
+    .select(['m.id', 'm.refId', 'm.objectType', 'm.refUpdatedAt', 'm.userId'])
+    .where('m.id', 'in', topIds)
+    .execute()
+
+  // Load objects via DataLoaders
+  const pageIds = metadata
+    .filter((m) => m.objectType === 'page')
+    .map((m) => Number(m.refId))
+    .filter((id) => !isNaN(id))
+
+  const templateIds = metadata.filter((m) => m.objectType === 'meetingTemplate').map((m) => m.refId)
+
+  const discussionIds = metadata
+    .filter((m) => m.objectType === 'retrospectiveDiscussionTopic')
+    .map((m) => m.refId)
+
+  const [pages, templates, discussions] = (await Promise.all([
+    pageIds.length > 0
+      ? (context as any).dataLoader.get('pages').loadMany(pageIds)
+      : Promise.resolve([]),
+    templateIds.length > 0
+      ? (context as any).dataLoader.get('meetingTemplates').loadMany(templateIds)
+      : Promise.resolve([]),
+    discussionIds.length > 0
+      ? (context as any).dataLoader.get('discussions').loadMany(discussionIds)
+      : Promise.resolve([])
+  ])) as [any[], any[], any[]]
+
+  // Map for easy access
+  const pageMap = new Map(
+    pages.filter((p: any) => p && !(p instanceof Error)).map((p: any) => [String(p.id), p])
+  )
+  const templateMap = new Map(
+    templates.filter((t: any) => t && !(t instanceof Error)).map((t: any) => [t.id, t])
+  )
+  const discussionMap = new Map(
+    (discussions as any[]).filter((d: any) => d && !(d instanceof Error)).map((d: any) => [d.id, d])
+  )
+
+  const rrgIds = Array.from(discussionMap.values())
+    .map((d: any) => d.discussionTopicId)
+    .filter(Boolean)
+
+  const rrgs =
+    rrgIds.length > 0
+      ? await (context as any).dataLoader.get('retroReflectionGroups').loadMany(rrgIds)
+      : ([] as any[])
+
+  const rrgMap = new Map(
+    rrgs.filter((r: any) => r && !(r instanceof Error)).map((r: any) => [r.id, r])
+  )
+
+  const MAX_RRF_SCORE = (keywordWeight + vectorWeight) * (1 / (k + 1))
+
+  const results = metadata
```

Comment:
-1 break this into a helper function. each objectType should have its own function